### ISSUE TITLE
Can't follow newly created users from Mastodon

### ISSUE THREAD

**Issue by tofuwabohu** [bookwyrm-social/bookwyrm#1928]
**Describe the bug**
I tried to follow a user on wyrms.de from various Mastodon instances. The profile is not restricted. It always showed that the account was followed but when refreshing the page, it shows "Cancel follow request". but the user doesn't get a notification that someone tried to follow them. There's no corresponding log in flower. 

I tried to reproduce that and found I can follow old users from Mastodon just fine but not new ones (there were no new users between August 2021 and this January). I can fetch their profiles to Mastodon but the Follow becomes a Follow Request that's never answered. Nothing in flower.

**To Reproduce**
Try to follow [@followTest@wyrms.de](http://wyrms.de/user/followTest) from any Mastodon instance.

**Additional context**
wyrms.de is on bw 0.3.0

**Comment by tofuwabohu** [comment]
Update: Apparently this is is bigger, the new users also can't follow users on other bookwyrm instances except bookwyrm,social. Following someone on kirja.casa looked like it worked but refreshing the profile just shows the "Follow" button again and that user doesn't appear in the following-list. The user has a protected account and accepted the request but wyrms.de didn't notice. The follow approach produced a flower log:
![image](https://user-images.githubusercontent.com/66644064/153212820-b9b22de1-09df-4b48-a3d3-6716a0b0e072.png)

Also, new users on wyrms.de do not receive dms from other instances and they don't produce a flower log. This works the other way round though, I can send dms from wyrms to masto and get the corresponding log in masto.

Summary: New users on wyrms.de can't follow users on other instances (masto and bw) and vice versa with bookwyrm.social being the exception (at least works sometimes). Dms seem to be working one way only.

**Comment by mouse-reeve** [comment]
Hm, this could be an issue with the public key verification -- that would reject activities without producing a flower error. Can you do `docker-compose logs --tail=100 web | grep 401` and see if that looks like it might be consistent? Sorry for not debugging this more directely for now -- I'm out of town this week and only online intermittently.

**Comment by tofuwabohu** [comment]
It doesn't but it shows other errors, thank you for answering even when you are abroad!
Maybe someone else has how to approach these PermissionDenied errors? There are plenty more like these. My first thought would be missing file system privileges but that doesn't make much sense since it only happens for newer users.
``` python
web_1             | Method Not Allowed: /inbox
web_1             | Method Not Allowed (GET): /user/<censored-username>/inbox
web_1             | Method Not Allowed: /user/<censored-username>/inbox
web_1             | Not Found: /user/followTest/status/7117/replies
web_1             | Method Not Allowed (GET): /inbox
web_1             | Method Not Allowed: /inbox
web_1             | Method Not Allowed (GET): /inbox
web_1             | Method Not Allowed: /inbox
web_1             | Method Not Allowed (GET): /inbox
web_1             | Method Not Allowed: /inbox

eb_1             | Method Not Allowed (GET): /inbox
web_1             | Method Not Allowed: /inbox
web_1             | Forbidden (Permission denied): /inbox
web_1             | Traceback (most recent call last):
web_1             |   File "/usr/local/lib/python3.9/site-packages/django/core/handlers/exception.py", line 47, in inner
web_1             |     response = get_response(request)
web_1             |   File "/usr/local/lib/python3.9/site-packages/django/core/handlers/base.py", line 181, in _get_response
web_1             |     response = wrapped_callback(request, *callback_args, **callback_kwargs)
web_1             |   File "/usr/local/lib/python3.9/site-packages/django/views/generic/base.py", line 70, in view
web_1             |     return self.dispatch(request, *args, **kwargs)
web_1             |   File "/usr/local/lib/python3.9/site-packages/django/utils/decorators.py", line 43, in _wrapper
web_1             |     return bound_method(*args, **kwargs)
web_1             |   File "/usr/local/lib/python3.9/site-packages/django/views/decorators/csrf.py", line 54, in wrapped_view
web_1             |     return view_func(*args, **kwargs)
web_1             |   File "/usr/local/lib/python3.9/site-packages/django/views/generic/base.py", line 98, in dispatch
web_1             |     return handler(request, *args, **kwargs)
web_1             |   File "/app/bookwyrm/views/inbox.py", line 45, in post
web_1             |     raise_is_blocked_activity(activity_json)
web_1             |   File "/app/bookwyrm/views/inbox.py", line 94, in raise_is_blocked_activity
web_1             |     raise PermissionDenied()
web_1             | django.core.exceptions.PermissionDenied
```

**Comment by mouse-reeve** [comment]
That PermissionDenied error from `raise_is_blocked_activity` should happen when the actor exists in your database but in a deleted state (either the account was deleted by the user, or bocked/deactivated by you). I might be on the wrong track about it being an authentication issue. I wonder if this is related to #1867 

Doing some testing between my dev instance and b.s, I was able to follow and receive follows between my dev instance and both bookwyrm.social and friend.camp, so it isn't as straightforward as a code error.

**Comment by mouse-reeve** [comment]
One thing that changed recently is the `./bw-dev update` command, and it's possible that it's not getting celery to pick up code changes correctly. If you haven't already, try `docker-compose restart celery_worker` just to make sure it's in sync with the app

**Comment by mouse-reeve** [comment]
Hm, you should have access to the `/admin` view with your regular user credentials, that is very odd. It's still working for me. If you're trying to access it as the test user, that might be the issue, or if you removed `is_superuser` or `is_staff` from your admin account? I'm not sure if that's related to the other issues, though.

**Comment by mouse-reeve** [comment]
When you say you can follow old users but not new ones, do you mean users that creates their accounts in the past versus ones that created their accounts recently, or ones that you added to your instance in the past versus recently?

**Comment by tofuwabohu** [comment]
Users who created their accounts on wyrms recently (new wyrms) be followed across instances. They can write dms to people across instances and follow them, though:
![image](https://user-images.githubusercontent.com/66644064/154142613-1bc60371-53f2-4afa-9c15-f5e174f3d3eb.png)

Answers to them do not come through, even if another, working account is mentioned. 
Masto -> new wyrm -> does not come through, no flower log
Masto -> old wyrm -> comes through, produces log and notification
Masto -> and and new wyrm at the same time -> does not come through, error in flower:

![image](https://user-images.githubusercontent.com/66644064/154143536-65136ee3-d9c5-4363-a3e8-0619582d0839.png)

Following from new wyrms to Masto works, but following users on other bw instances does not. I tried to follow you with followTest and it produced a flower log without error, but refreshing your profile sets back the follow button.
![image](https://user-images.githubusercontent.com/66644064/154143990-8c23cd36-dc3f-4b87-8fc6-0912da368221.png)

Am I missing any interactions to test?

Old wyrms can do everything as expected.

**Comment by mouse-reeve** [comment]
What does the error in flower say? Does it provide a stack trace?

**Comment by tofuwabohu** [comment]
``` python
Traceback (most recent call last):
  File "/app/bookwyrm/activitypub/base_activity.py", line 269, in resolve_remote_id
    data = get_data(remote_id)
  File "/app/bookwyrm/connectors/abstract_connector.py", line 277, in get_data
    raise ConnectorException()
bookwyrm.connectors.connector_manager.ConnectorException

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/celery/app/trace.py", line 451, in trace_task
    R = retval = fun(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/celery/app/trace.py", line 734, in __protected_call__
    return self.run(*args, **kwargs)
  File "/app/bookwyrm/views/inbox.py", line 109, in activity_task
    activity.action()
  File "/app/bookwyrm/activitypub/verbs.py", line 22, in action
    self.object.to_model()
  File "/app/bookwyrm/activitypub/base_activity.py", line 121, in to_model
    changed = field.set_field_from_activity(
  File "/app/bookwyrm/models/fields.py", line 78, in set_field_from_activity
    formatted = self.field_from_activity(value)
  File "/app/bookwyrm/models/fields.py", line 161, in field_from_activity
    return activitypub.resolve_remote_id(value, model=related_model)
  File "/app/bookwyrm/activitypub/base_activity.py", line 271, in resolve_remote_id
    raise ActivitySerializerError(
bookwyrm.activitypub.base_activity.ActivitySerializerError: Could not connect to host for remote_id: https://subversive.zone/users/tofuwabohu/statuses/107803866408227088
```
I'll try from another Mastodon instance as well in a sec

**Comment by tofuwabohu** [comment]
Oh interesting, mentioning both works from another Mastodon instance:
![image](https://user-images.githubusercontent.com/66644064/154145557-c5eace6a-0f1e-4450-add7-22df3463f2a7.png)

Only mentioning the new wyrm does not though and does not produce anyhting in flower.

**Comment by mouse-reeve** [comment]
The flower error looks like a transient error trying to connect to subversive.zone -- it could be a timeout on their end. 

Can you check the json view of the new user? it's at `/user/<username>.json`. I'd like to see if the public key is available normally, and if it looks different in any noticeable way to the older user's json view.

**Comment by tofuwabohu** [comment]
New [user](https://wyrms.de/user/followTest.json):
``` json
{"id": "http://wyrms.de/user/followTest",
 "type": "Person",
 "preferredUsername": "followTest",
 "inbox": "http://wyrms.de/user/followTest/inbox",
 "publicKey": {"id": "http://wyrms.de/user/followTest/#main-key",
 "owner": "http://wyrms.de/user/followTest",
 "publicKeyPem": "-----BEGIN PUBLIC KEY-----\nMIGfMA0GCSqGSIb3DQEBAQUAA4GNADCBiQKBgQCkkNWT23ZLKXZJrt8ANA5GqdDu\nwMHIps1zp8wRhzQBs52rglkuE0Ggdg/LsO9M4N5v6qWllaQQ53W2Kd4bOykuGyM2\nLYW4Pi7bsDHbzGz9rUpv02COemHwnNSti87AoPuIWvc7DrsYtDY+UlzjirexQ2w7\nQQhq/QwVzG8Opo9/PQIDAQAB\n-----END PUBLIC KEY-----"},
 "followers": "http://wyrms.de/user/followTest/followers",
 "following": "http://wyrms.de/user/followTest/following",
 "outbox": "http://wyrms.de/user/followTest/outbox",
 "endpoints": {"sharedInbox": "http://wyrms.de/inbox"},
 "icon": {},
 "bookwyrmUser": true,
 "manuallyApprovesFollowers": false,
 "discoverable": false,
 "@context": ["https://www.w3.org/ns/activitystreams", "https://w3id.org/security/v1",
 {"manuallyApprovesFollowers": "as:manuallyApprovesFollowers",
 "schema": "http://schema.org#",
 "PropertyValue": "schema:PropertyValue",
 "value": "schema:value"}]}
```

For some reason, all the references refer to http instead of https - I guess that could be a reason?

**Comment by mouse-reeve** [comment]
Yes, that could be the reason. Did you change the `USE_HTTPS` variable in .env? That should stay set to true if your instance loads via https (even if you don't use certbot)

**Comment by tofuwabohu** [comment]
I do not have that variable, I'll compare with the .env in the repo and include it, then build new and try with another user!
Should I change `DOMAIN=wyrms.de` to https://wyrms.de as well?

**Comment by mouse-reeve** [comment]
No, the `DOMAIN` is correct as is. You probably also need to edit the user's remote fields in the django admin to correct the http to https 🤔

**Comment by tofuwabohu** [comment]
It worked!! Thank you so much again. :) I created a new account and could follow it from Masto.

I did not remove USE_HTTPS conciously for sure, but either I messed up in vim or it was introduced after I set up my instance and then not being used for a while?

I'll now have a look on how to get access to the django admin, if it doesn't work I'll do it via console, it's not a terrible lot of new users.
